File: server/models/position.py
from pydantic import BaseModel

from .constants import Piece, Color
from .move import Move, Movement

bvalues = [
    [0, 3, 0, 3, 0, 3, 0, 3],
    [2, 0, 2, 0, 2, 0, 2, 0],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [1, 0, 2, 0, 2, 0, 1, 0],
    [0, 2, 0, 4, 0, 4, 0, 2],
    [3, 0, 6, 0, 6, 0, 3, 0],
    [0, 5, 0, 8, 0, 8, 0, 5],
    [5, 0, 9, 0, 9, 0, 6, 0],
]
wvalues = [row[::-1] for row in bvalues[::-1]]

def steps(x, y, dx, dy):
    x += dx
    y += dy

    while x >= 0 and x <= 7 and y >= 0 and y <= 7:
        yield (x, y)
        x += dx
        y += dy

class Position(BaseModel):
    move: Move = None
    turn: Color = 'w'
    field: list[list[Piece]] = []

    @classmethod
    def start(self):
        position = Position()
        position.board = [
            ['-', 'c', '-', 'c', '-', 'c', '-', 'c'],
            ['c', '-', 'c', '-', 'c', '-', 'c', '-'],
            ['-', 'c', '-', 'c', '-', 'c', '-', 'c'],
            ['-', '-', '-', '-', '-', '-', '-', '-'],
            ['-', '-', '-', '-', '-', '-', '-', '-'],
            ['C', '-', 'C', '-', 'C', '-', 'C', '-'],
            ['-', 'C', '-', 'C', '-', 'C', '-', 'C'],
            ['C', '-', 'C', '-', 'C', '-', 'C', '-'],
        ]
        position.turn = 'w'
        return position

    def possiblemoves(self):
        dirs = ((-1, -1), (-1, 1), (1, -1), (1, 1))
        moves = []
        for y, row in enumerate(self.field):
            for x, pc in enumerate(row):
                if pc == '-':
                    continue
                for dx, dy in dirs:
                    moves += self.getmoves(x, y, dx, dy)

        return moves
    
    def children(self):
        moves = self.possiblemoves()
        return [self.makemove(move) for move in moves]
    
    @property
    def value(self):
        result = 0
        for y, row in enumerate(self.field):
            for x, pc in enumerate(row):
                if pc == '-':
                    continue
                coef = 10 if pc == 'q' or pc == 'Q' else 1
                if pc == 'C' or pc == 'Q':
                    result += coef * wvalues[y][x]
                else:
                    result -= coef * bvalues[y][x]

        return result
    
    @property
    def bottom2(self):
        return min(1000000000, *[p.value for p in self.children()])

    @property
    def top2(self):
        return max(-1000000000, *[p.value for p in self.children()]);


    @property
    def board(self):
        return self.field
    
    @board.setter
    def board(self, board):
        self.field = board

    def makemove(self, move: Move):
        # TODO check against children
        
        field = [row[:] for row in self.field]
        for mv in move.movements:
            x0 = mv.srcx
            y0 = mv.srcy
            x1 = mv.dstx
            y1 = mv.dsty
            piece = 'q' if mv.piece == 'c' and y1 == 7 else ('Q' if mv.piece == 'C' and y1 == 0 else mv.piece)
            field[y0][x0] = '-'
            field[y1][x1] = piece
            if mv.take:
                stepx = 1 if x1 > x0 else -1
                stepy = 1 if y1 > y0 else -1
                i = 1
                while x0 + stepx * i != x1:
                    field[y0 + stepy * i][x0 + stepx * i] = '-'
                    i += 1

        turn = 'b' if self.turn == 'w' else 'w'

        position = Position(move = move, turn = turn)
        position.board = field

        return position
    
    def getmoves(self, x, y, dx, dy):
        moves = []
        if x + dx < 0 or x + dx > 7 or y + dy < 0 or y + dy > 7:
            return moves

        pc = self.field[y][x]
        if pc in ('c', 'q') and self.turn == 'w' or pc in ('C', 'Q') and self.turn == 'b' or pc == 'c' and dy <= 0 or  pc == 'C' and dy >= 0:
            return moves

        enemies = ('C', 'Q') if pc in ('c', 'q') else ('c', 'q')
        friends = ('c', 'q') if pc in ('c', 'q') else ('C', 'Q')

        if pc == 'c' or pc == 'C':
            if self.field[y + dy][x + dx] == '-':
                moves.append(Move(movements = [
                    Movement(piece = pc, srcx = x, srcy = y, dstx = x + dx, dsty = y + dy, take = '-')
                ]))
    
            if x + dx * 2 >= 0 and x + dx * 2 <= 7 and y + dy * 2 >= 0 and y + dy * 2 <= 7:
                if self.field[y + dy][x + dx] in enemies and self.field[y + dy * 2][x + dx * 2] == '-':
                    moves.append(Move(movements = [
                        Movement(piece = pc, srcx = x, srcy = y, dstx = x + dx * 2, dsty = y + dy * 2, take = self.field[y + dy][x + dx])
                    ]))

        if pc == 'Q' or pc == 'q':
            taken = False
            i = 1
            while True:
                if x + dx * i < 0 or x + dx * i > 7 or y + dy * i < 0 or y + dy * i > 7:
                    break
                if self.field[y + dy * i][x + dx * i] in friends:
                    break

                if self.field[y + dy * i][x + dx * i] == '-':
                    moves.append(Move(movements = [
                        Movement(piece = pc, srcx = x, srcy = y, dstx = x + dx * i, dsty = y + dy * i, take = '-')
                    ]))

                if x + dx * (i + 1) < 0 or x + dx * (i + 1) > 7 or y + dy * (i + 1) < 0 or y + dy * (i + 1) > 7:
                    break

                if not taken and self.field[y + dy * i][x + dx * i] in enemies and self.field[y + dy * (i + 1)][x + dx * (i + 1)] == '-':
                    taken = True
                    moves.append(Move(movements = [
                        Movement(piece = pc, srcx = x, srcy = y, dstx = x + dx * (i + 1), dsty = y + dy * (i + 1), take = self.field[y + dy * i][x + dx * i])
                    ]))

                i += 1

        return moves

File: client/src/models/position.ts
import {Piece, Color} from './constants';
import {Move, Movement} from './move';

export default class Position {
    public move: Move;
    public turn: Color = 'w';
    private field: Piece[][] = [];

    public static start() {
        let position = new Position(undefined, 'w', [
            ['-', 'c', '-', 'c', '-', 'c', '-', 'c'],
            ['c', '-', 'c', '-', 'c', '-', 'c', '-'],
            ['-', 'c', '-', 'c', '-', 'c', '-', 'c'],
            ['-', '-', '-', '-', '-', '-', '-', '-'],
            ['-', '-', '-', '-', '-', '-', '-', '-'],
            ['C', '-', 'C', '-', 'C', '-', 'C', '-'],
            ['-', 'C', '-', 'C', '-', 'C', '-', 'C'],
            ['C', '-', 'C', '-', 'C', '-', 'C', '-'],
        ]);
        return position;
    }

    public static parse(json) {
        return new Position(Move.parse(json.move), json.turn, json.field);
    }

    constructor(move?: Move, turn?: Color, field?: Piece[][]) {
        this.move = move || undefined;
        this.field = field || [];
        this.turn = turn || 'w';
    }

    public get board() {
        return this.field;
    }

    public get children() {
        let dirs = [[-1, -1], [-1, 1], [1, -1], [1, 1]];
        let moves = [];
        this.field.forEach((row, y) => {
            row.forEach((pc, x) => {
                if (pc === '-') {
                    return;
                }
                dirs.forEach(([dx, dy]) => moves.push(...this.getmoves(x, y, dx, dy)));
            });
        });
        return moves;
    }

    public pieceAt(x: number, y: number) {
        return this.field[y][x];
    }

    public makeMove(move: Move) {
        // TODO check against children
        
        let field = this.field.map(row => row.slice());
        move.movements.forEach(mv => {
            let x0 = mv.srcx;
            let y0 = mv.srcy;
            let x1 = mv.dstx;
            let y1 = mv.dsty;
            let piece = mv.piece === 'c' && y1 === 7 ? 'q' : (mv.piece === 'C' && y1 === 0 ? 'Q' : mv.piece);
            field[y0][x0] = '-';
            field[y1][x1] = piece;
            if (mv.take) {
                let stepx = Math.sign(x1 - x0);
                let stepy = Math.sign(y1 - y0);
                for (let i = 1; x0 + stepx * i !== x1; i ++) {
                    field[y0 + stepy * i][x0 + stepx * i] = '-';
                }
            }
        });

        let turn: Color = this.turn === 'w' ? 'b' : 'w';

        return new Position(move, turn, field);
    }

    private getmoves(x, y, dx, dy) {
        let moves = []
        if (x + dx < 0 || x + dx > 7 || y + dy < 0 || y + dy > 7) {
            return moves;
        }


        let pc = this.field[y][x];
        let enemies = ['c', 'q'].includes(pc) ? ['C', 'Q'] : ['c', 'q'];
        let friends = ['c', 'q'].includes(pc) ? ['c', 'q'] : ['C', 'Q'];
        if ((pc === 'c' || pc === 'q') && this.turn === 'w' ||
            (pc === 'C' || pc === 'Q') && this.turn === 'b' ||
            (pc === 'c' && dy <= 0 || pc === 'C' && dy >= 0)) {
            return moves;
        }

        if (pc === 'c' || pc === 'C') {
            if (this.field[y + dy][x + dx] === '-') {
                moves.push(new Move([new Movement(pc, x, y, x + dx, y + dy)]));
            }
    
            if (enemies.includes(this.field[y + dy][x + dx]) && this.field[y + dy * 2][x + dx * 2] === '-') {
                moves.push(new Move([new Movement(pc, x, y, x + dx * 2, y + dy * 2, this.field[y + dy][x + dx])]));
            }
        }

        if (pc === 'Q' || pc === 'q') {
            let taken = false;
            for (let i = 1; ; i ++) {
                if (x + dx * i < 0 || x + dx * i > 7 || y + dy * i < 0 || y + dy * i > 7) {
                    break;
                }
                if (friends.includes(this.field[y + dy * i][x + dx * i])) {
                    break;
                }

                if (this.field[y + dy * i][x + dx * i] === '-') {
                    moves.push(new Move([new Movement(pc, x, y, x + dx * 1, y + dy * 1)]));
                }

                if (x + dx * (i + 1) < 0 || x + dx * (i + 1) > 7 || y + dy * (i + 1) < 0 || y + dy * (i + 1) > 7) {
                    break;
                }
                if (!taken && enemies.includes(this.field[y + dy * i][x + dx * i]) && this.field[y + dy * (i + 1)][x + dx * (i + 1)] === '-') {
                    taken = true;
                    moves.push(new Move([new Movement(pc, x, y, x + dx * (i + 1), y + dy * (i + 1), this.field[y + dy * i][x + dx * i])]));
                }
            }
        }

        return moves;
    }
}

